File: model/src/main/java/io/github/chriszhong/model/organization/event/InstanceGoalEventFactory.java
package io.github.chriszhong.model.organization.event;

import io.github.chriszhong.model.organization.entity.InstanceGoal;

/**
 * The {@linkplain InstanceGoalEventFactory} interface defines the API for constructing {@linkplain InstanceGoalEvent}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface InstanceGoalEventFactory {

	/**
	 * Constructs an {@linkplain InstanceGoalEvent}.
	 *
	 * @param category
	 *            the {@linkplain EventCategory}.
	 * @param goal
	 *            the {@linkplain InstanceGoal}.
	 *
	 * @return a {@linkplain InstanceGoalEvent}.
	 */
	InstanceGoalEvent build(EventCategory category, InstanceGoal goal);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/InstanceGoalEventFactory.kt
package runtimemodels.chazm.model.entity

import runtimemodels.chazm.api.entity.InstanceGoal
import runtimemodels.chazm.model.event.EventType

/**
 * The [InstanceGoalEventFactory] interface defines the API for constructing [InstanceGoalEvent]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface InstanceGoalEventFactory {
    /**
     * Constructs an [InstanceGoalEvent].
     *
     * @param category the [EventType].
     * @param goal     the [InstanceGoal].
     * @return a [InstanceGoalEvent].
     */
    fun build(category: EventType, goal: InstanceGoal): InstanceGoalEvent
}
